Add bitcoin address label to request payment QR code

makevoid opened the pull request:
This is a change for the QT wallet

**Why this change?**
This way the user can be sure that the QR code scanner app he/she's using is not displaying a different address to maliciously steal the funds.

I've seen an android app in the wild that is a simple qr code scanner but when it scans a bitcoin address it replaces it with a different address belonging to the hacker. I'm starting to get worried of bitcoin address qr codes without a label that states the address in clear so that the user can always check that he/she is depositing the funds to the desired address. 
That's why I decided to modify the bitcoin qt code.

---

**Code changes:**
- renamed myImage to qrImage
- added qrAddrImage in which the bitcoin address label is painted onto
- scaled the saved pixmap with antialiasing to improve readability of the text and the scanning of the qrcode

makevoid committed:
Add address label to request payment QR Code (QT)

In the Receive 'Tab' of the QT wallet, when 'Show'ing a previously requested payment, add a label underneath the QR Code showing the bitcoin address where the funds will go to.

This way the user can be sure that the QR code scanner app the user using is reading the correct bitcoin address, preventing funds to be stolen.

Includes fix for HiDPI screens by @jonasschnelli.

makevoid commented:
That's the preview of the change, with the saved image file antialiased:
![](http://mkvd.s3.amazonaws.com/pics/qr_bc_core_pr.png)

laanwj commented:
Haven't looked at the code, but screenshot looks good to me, nice work.

jonasschnelli commented:
Thanks!

Having a malicious QRCode can probably one happen if you system, Bitcoin-Core or at least libqrencode is compromise.

You PR would only protect from an attack on libqrencode and only if the user uses a self-compiled Bitcoin-Core (official binaries are static linked). Because, if you could tamper Bitcoin-Core, you could change the address at a core point which would also provide a malicious address below the QRCode.

This change would def. make sense if we would have HD for key derivation.
You could use a smartphone with the same xpub master key and verify the address at given keypath by comparing QRCodes or addresses together with the keypath.

IMO this PR does not hurt, although I don't believe that is increases security.

UtACK.

laanwj commented:
> IMO this PR does not hurt, although I don't believe that is increases security.

If anything at the printer side is compromised, you're out of luck. But what about a malicious QR code *reader*? I think it could help there, comparing the address before accepting the transaction in your wallet.

MarcoFalke commented:
Right, I think it makes sense to have a human readable version of the QR code besides it. Concept ACK.

jonasschnelli commented:
Yes. I just thought people could already do this because they have the address below the QRCode already. But this PR makes it more prominent available.

paveljanik commented:
Testing on OS X. The image is blurred. Original image is OK:

Current master: 
![before](https://cloud.githubusercontent.com/assets/6848764/13494701/2c81a37c-e146-11e5-8cb0-46b445e491db.png)

This PR:
![after](https://cloud.githubusercontent.com/assets/6848764/13494704/32154172-e146-11e5-976b-c1773a7283d2.png)

The Request payment window shows only the top part of the address:

<img width="493" alt="screen shot 2016-03-03 at 13 43 03" src="https://cloud.githubusercontent.com/assets/6848764/13494723/512705d2-e146-11e5-9f9d-9cb438736507.png">

Can you fix these problems?

jonasschnelli commented on src/qt/receiverequestdialog.cpp:
```diff
@@ -45,7 +45,7 @@ QImage QRImageWidget::exportImage()
     if(!pixmap())
         return QImage();
-    return pixmap()->toImage().scaled(EXPORT_IMAGE_SIZE, EXPORT_IMAGE_SIZE);
+    return pixmap()->toImage().scaled(EXPORT_IMAGE_SIZE, EXPORT_IMAGE_SIZE, Qt::KeepAspectRatio, Qt::SmoothTransformation);
```
QRCodes should not use anti-aliasing (QT::SmoothTransformation). I guess this should be removed.

makevoid commented:
I removed the anti-aliasing (I should I've imagined that it was bad for qr code scanning) and now it shouldn't have that blurring problem on OSX as it uses the same size as the size it's using to display it (`300`), I will double-check it in few hours.

The only thing that's changed now is the export size, the final image is 44 pixel larger for each dimension. The good thing is now I can reuse the constant `EXPORT_IMAGE_SIZE` (now changed to `QR_IMAGE_SIZE`, used only in this file), also there's no need to re-scale the image again and again anymore.

I can squash the three commits to a single one and force push to this branch when needed.

paveljanik commented:
One new warning generated:

```
+qt/receiverequestdialog.cpp:184:73: warning: implicit conversion from 'double' to 'int' changes value from 8.5 to 8 [-Wliteral-conversion]
+            painter.setFont(QFont(GUIUtil::fixedPitchFont().toString(), 8.5));
+                            ~~~~~                                       ^~~
+1 warning generated.
```

luke-jr commented:
So now the window displays the address twice? Maybe just add it to the image when it's being saved...?

paveljanik commented:
@luke-jr If displayed properly, the address is displayed four times! Title bar, image, bitcoin: URL and address. 8)

jonasschnelli commented:
Hmm... font feels a bit small on my end (HiDPI/retina OSX):

<img width="599" alt="bildschirmfoto 2016-03-03 um 17 06 04" src="https://cloud.githubusercontent.com/assets/178464/13500300/44c1a77c-e162-11e5-94eb-13a8cd99b465.png">

jonasschnelli commented on src/qt/receiverequestdialog.cpp:
```diff
@@ -166,20 +166,27 @@ void ReceiveRequestDialog::update()
-            ui->lblQRCode->setPixmap(QPixmap::fromImage(myImage).scaled(300, 300));
+            QImage qrAddrImage = qrImage.scaled(QR_IMAGE_SIZE, QR_IMAGE_SIZE);
+            QPainter painter(&qrAddrImage);
+            painter.setFont(QFont(GUIUtil::fixedPitchFont().toString(), 8));
```
Maybe add a [`QFont::setPointSize()`](http://doc.qt.io/qt-5/qfont.html#setPointSize) here to avoid HiDPI differences?

laanwj commented:
> So now the window displays the address twice? Maybe just add it to the image when it's being saved...?

Well at least people can't complain they're missing the address :-)
I don't think this is much of a concern, it's not like the window is short on space.

laanwj commented:
One thing I do wonder about though: do QR scanners need an empty border around a QR code? If so, how large should it be?

jonasschnelli commented:
@laanwj: I recently played around with max QR size/bytes, black/white (inverted) and the white-space-border. I could not find any specification about the "quite zone" (white border), but I guess "one square width" as min border size should be more as sufficient. I guess it depends on the QRCode reader implementation. But I could not read the QRCode with common readers if there where no white borders.

sipa commented:
Scanners don't need a border I believe, but printers may.

sipa commented:
Seems I misremembered. Some googling shows that the spec actually includes
a quiet zone around the QR code, 4 modules wide, with the same color as the
white modules.

paveljanik commented:
Current status:

<img width="479" alt="screen shot 2016-03-10 at 13 50 18" src="https://cloud.githubusercontent.com/assets/6848764/13669936/1b7fbb1e-e6c7-11e5-9447-73cd380ddcbb.png">

jonasschnelli commented:
Still not really readable on retina/HiDPI OSX:
<img width="663" alt="bildschirmfoto 2016-03-13 um 10 24 24" src="https://cloud.githubusercontent.com/assets/178464/13727910/d85aee4e-e905-11e5-8cf3-6a57df683fb8.png">

jonasschnelli commented on src/qt/receiverequestdialog.cpp:
```diff
@@ -166,20 +166,30 @@ void ReceiveRequestDialog::update()
+            painter.setFont(QFont(GUIUtil::fixedPitchFont().toString(), 8));
+            QRect paddedRect = qrAddrImage.rect();
+            paddedRect.setHeight(QR_IMAGE_SIZE+12);
+            painter.drawText(paddedRect, Qt::AlignBottom|Qt::AlignCenter, info.address);
```
Maybe some clever rule to detect the width of the text we want to draw – and make it bigger if there is more horizontal space?

Get width:
```c++
 QFontMetrics fm(painter.font());
 int pixelsWide = fm.width(info.address);
```

jonasschnelli commented:
ping @makevoid: I think this PR is almost ready. We just need a way that the font size respects the screen density (HiDPI).

jonasschnelli commented:
ping @makevoid

laanwj commented:
> We just need a way that the font size respects the screen density (HiDPI).

I'd say no because this is exported to an image file. We don't want the output to be dependent on the user's screen resolution.

jonasschnelli commented:
@laanwj: check https://github.com/bitcoin/bitcoin/pull/7636#issuecomment-195920098
The font size on HiDPI is different then on non-HiDPI.

I can take care about this and finish this PR.

makevoid commented:
Sorry but I don't have an HiDPI screen to test with yet, is there anything I can do? should I squash the commits?

laanwj commented:
> The font size on HiDPI is different then on non-HiDPI.

Just a matter of misunderstanding / wrong formulation then - what we want here is to use a font size that is fixed in pixel size, just like the QR code itself is.

jonasschnelli commented:
The QRCode is not rendered HiDPI (not required). Setting the font size in pixel works (because the size of the `QImage` is fixed in pixels not in points).
The only little downside is the missing HiDPI support for the new address (looks a bit pixled here).

<img width="599" alt="bildschirmfoto 2016-06-13 um 15 41 25" src="https://cloud.githubusercontent.com/assets/178464/16009555/859230ee-317d-11e6-9748-e989b89346cf.png">
<img width="518" alt="bildschirmfoto 2016-06-13 um 15 40 57" src="https://cloud.githubusercontent.com/assets/178464/16009556/85b892fc-317d-11e6-8f43-14650d753ba7.png">


Here's the diff.
@makevoid: Can you apply and squash?
```diff
diff --git a/src/qt/receiverequestdialog.cpp b/src/qt/receiverequestdialog.cpp
index 1e104bb..b81a039 100644
--- a/src/qt/receiverequestdialog.cpp
+++ b/src/qt/receiverequestdialog.cpp
@@ -183,7 +183,9 @@ void ReceiveRequestDialog::update()
             qrAddrImage.fill(0xffffff);
             QPainter painter(&qrAddrImage);
             painter.drawImage(0, 0, qrImage.scaled(QR_IMAGE_SIZE, QR_IMAGE_SIZE));
-            painter.setFont(QFont(GUIUtil::fixedPitchFont().toString(), 8));
+            QFont font = QFont(GUIUtil::fixedPitchFont().toString());
+            font.setPixelSize(12);
+            painter.setFont(font);
             QRect paddedRect = qrAddrImage.rect();
             paddedRect.setHeight(QR_IMAGE_SIZE+12);
             painter.drawText(paddedRect, Qt::AlignBottom|Qt::AlignCenter, info.address);
```

MarcoFalke commented:
> `+            QFont font = QFont(GUIUtil::fixedPitchFont().toString());`

fixedPitchFont() already returns QFont. Is this odd constructor necessary?

jonasschnelli commented:
@MarcoFalke: Oh yes.
@makevoid: please use `QFont font = GUIUtil::fixedPitchFont();`.

makevoid commented:
@jonasschnelli done: https://github.com/bitcoin/bitcoin/pull/7636/commits/1c2a1bac0ad6901868c1a4003c7cbddcdf46a29b#diff-23ecec8032dd36794001112346601b6aR186

MarcoFalke commented:
utACK 1c2a1ba

jonasschnelli commented:
Tested ACK 1c2a1bac0ad6901868c1a4003c7cbddcdf46a29b.

Looks also good on Win(10).
HiDPI issues on windows are unrelated to this PR (we need Qt5.6.1).

<img width="436" alt="bildschirmfoto 2016-06-14 um 13 03 32" src="https://cloud.githubusercontent.com/assets/178464/16040721/b1c0b92e-3231-11e6-97de-9755a20440f5.png">
